src/Plugin/QuickForm/SaatenUnionSpraying.php

```diff
@@ -0,0 +1,746 @@
+    $render['label'] = $config['label'];
+
+    // Save units to a variable for now.
+    // The key may be saved as units or units_id.
+    $units_key_name = 'units';
+    $units = $config['units'];
+
+    // Check if unit options are provided.
+    if (!empty($units['#options'])) {
+      $units_options = $units['#options'];
+
+      // If a numeric value is provided, assume these are term ids.
+      if (is_numeric(key($units_options))) {
+        $units_key_name = 'units_id';
+      }
+
+      // Render the units as select options.
+      $units += [
+        '#type' => 'select',
+        '#options' => $units_options,
+      ];
+
+      // If the unit value is hard-coded add a field suffix to the value field
+      // with the first option label.
+      if (isset($units['#value'])) {
+        $render['value']['#field_suffix'] = current($units_options);
+      }
+    }
+    // Else default to entity_autocomplete unit terms. Use the units_id key.
+    else {
+      $units_key_name = 'units_id';
+      // Add entity_autocomplete.
+      $units += [
+        '#type' => 'entity_autocomplete',
+        '#placeholder' => $this->t('Units'),
+        '#target_type' => 'taxonomy_term',
+        '#selection_handler' => 'default',
+        '#selection_settings' => [
+          'target_bundles' => ['unit'],
+        ],
+        '#tags' => FALSE,
+        '#size' => 15,
+      ];
+    }
+
+    // Include units in render array.
+    $render[$units_key_name] = $units;
+
+    // Check if the quantity is required.
+    if (!empty($config['required'])) {
+      $render['#required'] = TRUE;
+      $render['value']['#required'] = TRUE;
+    }
+
+    // Remove the border if needed.
+    if (empty($config['border'])) {
+      $render['#attributes']['class'][] = 'no-border';
+    }
+
+    return $render;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function prepareNotes(array $note_fields, FormStateInterface $form_state): array {
+    // Prepend additional note fields.
+    array_unshift(
+      $note_fields,
+      ...[
+        [
+          'key' => 'end_date',
+          'label' => $this->t('End Date'),
+        ],
+        [
+          'key' => 'justification_target',
+          'label' => $this->t('Justification/Target'),
+        ],
+        [
+          'key' => 'plant_growth_stage',
+          'label' => $this->t('Plant Growth Stage'),
+        ],
+        [
+          'key' => 'weather',
+          'label' => $this->t('Weather'),
+        ],
+        [
+          'key' => 'wind_direction',
+          'label' => $this->t('Wind direction'),
+        ],
+        [
+          'key' => 'equipment_rinsed',
+          'label' => $this->t('Equipment triple-rinsed'),
+        ],
+        [
+          'key' => 'equipment_clear',
+          'label' => $this->t('Equipment all clear'),
+        ],
+        [
+          'key' => 'equipment_washed',
+          'label' => $this->t('Equipment clear washed'),
+        ],
+      ]
+    );
```

Comment:
You can move these into the `submitForm()` method where this function is called. That would be better and help keep this function abstract in case others would like to reference it. The note fields look great!

Reply:
I have moved this object to the submit function